lib/src/machine.rs

```diff
@@ -0,0 +1,379 @@
+use crate::{
+    ast::{Environment, Instruction, LangValue, Program, StackIdentifier},
+    error::RuntimeError,
+};
+use failure::Fallible;
+use std::iter;
+
+/// Wrapper around a vec of stacks, to make it a bit easier to initialize/use.
+///
+/// All stack manipulation logic should be implemented here, to keep it
+/// contained and scalable.
+/// This is zero-cost at runtime, see [newtype pattern](https://doc.rust-lang.org/1.0.0/style/features/types/newtype.html).
+#[derive(Debug, PartialEq)]
+pub struct Stacks(Vec<Vec<LangValue>>);
+
+impl Stacks {
+    fn new(env: &Environment) -> Self {
+        Self(
+            iter::repeat(0)
```

Comment:
why the repeat(0)?

Reply:
What I'm trying to do here is a create a `Vec` of `Vec`s, where the length of the parent vec is `env.num_stacks` (and each sub-vec is one of the user's stacks). `iter::repeat` repeats the given value infinitely, so together `iter::repeat(0).take(env.num_stacks)` creates an iter that repeats `0` by `env.num_stacks` times. The `0` isn't important, I just picked a value. Once it has that, it generates one new vec per element. As I'm typing this though, I realized there's a slightly cleaner way to do this, so I'm going to update it and add a comment